build.py

```diff
@@ -834,12 +842,46 @@ def TEMPFILE(cmd_, comstr=None):
             if list_targets[0].startswith("@"):
                 test_name = list_targets[0]
                 test_name = '+' + test_name[1:]
+
+                # These take priority over unit test shortcut names.
+                self.unittest_skipped_shortcuts.add(test_name)
+                if test_name in self.unittest_shortcuts.keys():
+                    del self.unittest_shortcuts[test_name]
+
                 self.builds.append(dict(
                         rule='RUN_TEST',
                         outputs=test_name,
                         inputs=list_sources
                     ))
 
+                # Add shortcuts for this unit test.
+                for unit_test_source_file in \
+                        n.executor.get_all_children()[0].executor.get_all_sources():
+                    # Get the file name
+                    pos_last_slash = str(unit_test_source_file).rfind(os.path.sep)
+                    # Strip the '.o' or '.obj'
+                    stripped_name = str(unit_test_source_file)[pos_last_slash + 1:]
+                    pos_last_dot = stripped_name.find(".")
+                    stripped_name = stripped_name[:pos_last_dot]
+
+                    if "_test" in stripped_name[len(stripped_name) - 5:]:
+                        # Add suffix to tests on Windows to match other unit tests
+                        suffix = ".exe" if self.globalEnv.TargetOSIs('windows') else ""
+                        stripped_name = '+' + stripped_name + suffix
+                        if (stripped_name not in self.unittest_shortcuts and stripped_name not
+                                in self.unittest_skipped_shortcuts):
+                            # Add a shortcut for the given unit test file name.
+                            self.unittest_shortcuts[stripped_name] = dict(
+                                rule='RUN_TEST',
```

Comment:
I would instead make these aliases of the test runs. Then if you happened to do `ninja +foo_test_file_a +foo_test_file_b` then it will only run `+foo_test` once.  This also makes it easier to handle cases where the same file basename is part of multiple tests since you could just just run both (although it may be better to error in that case so we can fix our file names).


Alternatively, if you are willing to modify the unittest lib slightly, you could make the `TEST` family of macros record `__FILE__` and add a way to filter on basename. Then you could have RUN_TEST take an optional `$TEST_ARGS` variable that is expanded to a filter for each shortcut.

Reply:
Done, looks like the 'phony' rule in ninja does exactly what we want here.